src/dune_rules/menhir/menhir_rules.ml

```diff
@@ -316,6 +321,85 @@ module Run (P : PARAMS) = struct
 
   (* ------------------------------------------------------------------------ *)
 
+  (* [process_messages stanza m] generates rules to maintain a .messages file
+     corresponding to a Menhir grammar. *)
+
+  (* Given a Menhir stanza
+
+       (menhir (modules parser) (messages parser_messages))
+
+     the code below defines rules essentially equivalent to
+
+       (rule
+        (with-stodut-to parser_message.messages.list
+         (run menhir parser.mly --list-errors)))
+
+       (rule
+        (with-stdout-to parser_messages.messages.update
+         (run menhir parser.mly --update-errors parser_messages.messages)))
+
+       (rule
+        (with-stdout-to parser_messages.messages.new
+         (run menhir parser.mly --merge-errors parser_messages.messages.list
+                                --merge-errors parser_messages.messages.update)))
+
+       (rule
+        (target parser_messages.ml)
+        (action
+         (prog
+          (diff parser_messages.messages.new parser_messages.messages)
+          (run menhir parser.mly --compile-errors parser_messages.messages))))
+  *)
+
+  let messages m = Path.relative (Path.build dir) (m ^ ".messages")
+  let messages_new m = Path.Build.relative dir (m ^ ".messages.new")
+  let messages_list m = Path.Build.relative dir (m ^ ".messages.list")
+  let messages_update m = Path.Build.relative dir (m ^ ".messages.update")
+  let messages_ml m = Path.Build.relative dir (m ^ ".ml")
+
+  let process_messages (stanza : stanza) (m : string) : unit Memo.t =
+    let open Memo.O in
+    let* expanded_flags = expand_flags stanza.flags in
+    let menhir ?stdout_to args =
+      menhir
+        ?stdout_to
+        ([ Command.Args.dyn expanded_flags; Deps (sources stanza.modules) ] @ args)
+    in
+    let list_errors = menhir ~stdout_to:(messages_list m) [ A "--list-errors" ] in
+    let update_errors =
+      menhir ~stdout_to:(messages_update m) [ A "--update-errors"; Dep (messages m) ]
+    in
+    let merge_errors =
+      menhir
+        ~stdout_to:(messages_new m)
+        [ A "--merge-errors"
+        ; Dep (Path.build (messages_list m))
+        ; A "--merge-errors"
+        ; Dep (Path.build (messages_update m))
+        ]
+    in
+    let compile_errors =
+      let diff =
+        let action =
+          let open Action_builder.O in
+          let+ () = Action_builder.path (Path.build (messages_new m))
+          and+ () = Action_builder.path (messages m) in
+          Action.Full.make (Promote.Diff_action.diff (messages m) (messages_new m))
+        in
+        { With_targets.build = action; targets = Targets.empty }
+      in
+      let compile =
+        menhir ~stdout_to:(messages_ml m) [ A "--compile-errors"; Dep (messages m) ]
+      in
+      Action_builder.progn [ diff; compile ]
+    in
+    Memo.sequential_iter
```

Comment:
Maybe it makes sense to sandbox these by default?

Reply:
The calls to `menhir` all are already sanboxed: https://github.com/ocaml/dune/blob/95dc808fe6c746740449fd0118af0e047165aa4f/src/dune_rules/menhir/menhir_rules.ml#L82-L88